### PR TITLE
add detection for meta mask

### PR DESCRIPTION
**Pull request by H34D**
## Description

fix (hack) signing in meta mask and other providers

### CODE REVIEW

File: src/ocean/ServiceAgreements/ServiceAgreement.ts [hunk 1]
```diff
@@ -86,8 +86,14 @@ export default class ServiceAgreement extends OceanBase {
...
 
-        const serviceAgreementHashSignature = await Web3Provider
-            .getWeb3().eth.sign(serviceAgreementHash, consumer.getId())
+        let serviceAgreementHashSignature: string
+        const web3 = Web3Provider.getWeb3()
+        if (web3.currentProvider.isMetaMask) {
+            // password is injected by metamask, dont try to set it!
+            serviceAgreementHashSignature = await web3.eth.personal.sign(serviceAgreementHash, consumer.getId(), null)
+        } else {
+            serviceAgreementHashSignature = await web3.eth.sign(serviceAgreementHash, consumer.getId())
```
Comment (diminator) [inline src/ocean/ServiceAgreements/ServiceAgreement.ts]: not `personal.sign`? 

How would that play out with passwords (config, HDwallet, ...)

File: src/ocean/ServiceAgreements/ServiceAgreement.ts [hunk 2]
```diff
@@ -86,8 +86,14 @@ export default class ServiceAgreement extends OceanBase {
             valueHashes,
             timeoutValues)
 
-        const serviceAgreementHashSignature = await Web3Provider
-            .getWeb3().eth.sign(serviceAgreementHash, consumer.getId())
+        let serviceAgreementHashSignature: string
+        const web3 = Web3Provider.getWeb3()
+        if (web3.currentProvider.isMetaMask) {
+            // password is injected by metamask, dont try to set it!
+            serviceAgreementHashSignature = await web3.eth.personal.sign(serviceAgreementHash, consumer.getId(), null)
```
Comment (ssallam) [inline src/ocean/ServiceAgreements/ServiceAgreement.ts]: The third argument is not necessary, but it is ok.

Comment (H34D) [reply src/ocean/ServiceAgreements/ServiceAgreement.ts]: yes, it only there to make sure that this one needs a password

File: src/ocean/ServiceAgreements/ServiceAgreement.ts (hunk 1 above)
Comment (H34D) [reply src/ocean/ServiceAgreements/ServiceAgreement.ts]: it's `eth.personal.sign` https://web3js.readthedocs.io/en/1.0/web3-eth-personal.html#sign

Comment (H34D) [reply src/ocean/ServiceAgreements/ServiceAgreement.ts]: > How would that play out with passwords (config, HDwallet, ...)

we should really go for https://web3js.readthedocs.io/en/1.0/web3-eth-accounts.html#sign, that requires raw key material that can be produced with hdwallet/bip39

**Pull request merged**